### PR TITLE
Updates to 2024 RIKEN tutorial for DYAD

### PR DESCRIPTION
**Pull request by ilumsden**
This PR adds several updates to the work @vsoch has already done for the 2024 RIKEN tutorial for the DYAD component.

More specifically, it does the following:
* Updates the Docker stuff to build the newest version of DYAD
* Updates the Docker stuff to build the DLIO benchmark and obtain the PyTorch data loader for DYAD
* Updates the DYAD notebook using material from the HiCOMB and SC papers
* Updates the DYAD notebook to use a simple run of DLIO to show DYAD's benefits for distributed deep learning training (note: no GPUs are needed for this!)

### CODE REVIEW

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn [hunk 1]
```diff
@@ -101,6 +102,10 @@ CMD ["flux", "start", "--test-size=4", "jupyter", "lab"]
 # This won't be available in K8s, but will be for a single container build
 COPY ./tutorial /home/jovyan/flux-tutorial-2024
 
+USER root
+RUN mv /home/jovyan/dlio_extensions /home/jovyan/flux-tutorial-2024
```
Comment (vsoch) [inline 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:106]: Just want to double check the outcome of this - it puts dlio_extensions into flux-tturoial-2024? E.g.,

```
/home/jovyan/flux-tutorial-2024/dlio-extensions/
```
Any reason to not put the COPY command for the tutorial first and put the directory there in the same build step? It might get around needing to switch back and forth from the user again (but putting it all in the root block).

Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:106]: You know... that makes a lot more sense. Let me change that quickly.

Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn]: Fixed

Comment (vsoch) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn]: That looks good! Have you run this notebook locally (with the updated build) and can confirm everything is working as expected?

Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn]: I haven't run since moving the COPY and mv, but everything was working before that. I tested it personally, and someone else in our lab here at UT tested it too.

Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn]: To be clear, when I say "working", I mean that everything builds, and the Flux side runs. I need to finish the new stuff for the DYAD notebook before I can test if that stuff works w.r.t running

Comment (vsoch) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn]: okay, let's wait then for that, and one final run through that everything is working. Then we can merge into the PR branch. My main desire is not to merge something that has not been fully tested to be working.

Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn]: Sounds good. Baring anything unexpected, I should have that done sometime tomorrow.

File: 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/dyad_dlio.ipynb [hunk 2]
```diff
@@ -0,0 +1,345 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "dd3e912b-3428-4bc7-88bd-97686406b75a",
```
Comment (vsoch) [inline 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/dyad_dlio.ipynb:5]: What about the original dyad notebook (see it here https://github.com/flux-framework/Tutorials/tree/4adc40d7519d592b467eac23c04f421383089874/2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook) is this supplement to that or intended to replace it? A

Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/dyad_dlio.ipynb:5]: The `dyad_dlio.ipynb` is going to replace the existing `dyad.ipynb`. In fact, when I'm done editing it, I'm going to rename `dyad_dlio.ipynb` to `dyad.ipynb`. I'm just keeping the original around until I've finished the text in the new notebook.

Comment (vsoch) [reply 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/dyad_dlio.ipynb:5]: ok! Ping me when you are all done and ready for a review. 

Actually that's what draft is for right?

Right, right... :laughing:

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM fluxrm/flux-sched:jammy
+FROM fluxrm/flux-sched:focal
```
Comment (vsoch) [inline 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:1]: Why did you change to focal? Jammy is newer (and it's going to last longer so I tend to use it).

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn [hunk 4]
```diff
@@ -79,11 +101,17 @@ RUN python3 -m pip install jupyter_app_launcher && \
...
 
-# No permission errors here
-USER ${NB_USER}
+# Give jovyan user permissions to tutorial materials
+RUN chmod -R 777 ~/flux-tutorial-2024
+
 WORKDIR $HOME
 COPY ./docker/flux-icon.png $HOME/flux-icon.png
 
+# RUN ${VENV_SOURCE} && \
```
Comment (vsoch) [inline 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:110]: What is this commented out section?

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn [hunk 5]
```diff
@@ -96,11 +124,14 @@ COPY ./docker/entrypoint.sh /entrypoint.sh
 
 # This is for a local start
 COPY ./docker/start.sh /start.sh
-CMD ["flux", "start", "--test-size=4", "jupyter", "lab"]
 
-# This won't be available in K8s, but will be for a single container build
-COPY ./tutorial /home/jovyan/flux-tutorial-2024
+RUN mkdir -p $HOME/.local/share && \
+    chmod 777 $HOME/.local/share
```
Comment (vsoch) [inline 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:129]: What is going to go here?

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,54 @@
+FROM fluxrm/flux-sched:focal
```
Comment (vsoch) [inline Dockerfile:1]: What is this new Dockerfile?

**Review by vsoch — COMMENTED** [review]
I'll review it interactively, but a few questions first!

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn (hunk 3 above)
Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:1]: Because of requirements for some dependencies of DLIO (namely Tensorflow), we need certain packages that are not available on Jammy. @hariharan-devarajan could give you a better sense of what those packages are than I could.

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn (hunk 4 above)
Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:110]: Oh, I should remove that. It was stuff we had from trying to get DLIO to work.

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn (hunk 3 above)
Comment (vsoch) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:1]: okay that's reasonable - focal goes EOL in about a year, so I'm OK with one more year of tutorial! We will likely need to update it next year.

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn (hunk 5 above)
Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:129]: The diff seems to be a little confused there. Two things are happening:
1. The `COPY` directive was moved towards the top of the file
2. We added a `chmod` to `/home/jovyan/.local/share` because we were getting permission denied errors on that directory

File: Dockerfile (hunk 6 above)
Comment (ilumsden) [reply Dockerfile:1]: Oh, I meant to delete that. Similar to the comments, that was for testing.

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn [hunk 7]
```diff
@@ -9,32 +9,51 @@ USER root
 ENV NB_USER=jovyan \
     NB_UID=1000 \
     HOME=/home/jovyan
+    # VENV_DIR=/home/jovyan/.flux_tutorial_venv
```
Comment (vsoch) [inline 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:12]: Same here - can this commented out line be removed?

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn [hunk 8]
```diff
@@ -9,32 +9,51 @@ USER root
...
+    python3-pip \
+    python3-venv \
+    openmpi-bin \
+    openmpi-common \
+    libopenmpi-dev \
+    liblz4-dev \
+    tini \
+    # requirement for nbgitpuller
+    git
+#&& rm -rf /var/lib/apt/lists/*
```
Comment (vsoch) [inline 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:41]: Can this be uncommented to clean up extra files?

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn (hunk 7 above)
Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:12]: I've got a commit that will remove that

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn [hunk 9]
```diff
@@ -43,28 +62,31 @@ RUN git clone https://github.com/openucx/ucx.git \
...
-       --without-rocm --without-xpmem --without-fuse3 --without-ugni --prefix=/usr CC=$(which gcc) CXX=$(which g++) \
+    --without-go --without-java --disable-cma --without-cuda --without-gdrcopy --without-verbs --without-knem --without-rmdacm \
+    --without-rocm --without-xpmem --without-fuse3 --without-ugni --prefix=/usr CC=$(which gcc) CXX=$(which g++) \
     && make -j \
     && sudo make install \
     && cd .. \
     && rm -rf ucx
 
-RUN git clone https://github.com/TauferLab/dyad.git \
+# RUN $VENV_SOURCE \
```
Comment (vsoch) [inline 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:72]: This is commented out too.

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn (hunk 8 above)
Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:41]: Probably, although I haven't tested it since we got DLIO working. I can always uncomment it, and let the CI take a crack at building the image. If it works for the CI, it works for me

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn (hunk 5 above)
Comment (vsoch) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:129]: I think this is happening because you are running all of these as `USER root`. Running the add user command does not switch the user, it just adds them. So the commands to install (with sudo) if I'm reading this right are being run by root, and the COPY directives would have root permissions too. If you want the jovyan user to have ownership of that space I would also COPY, etc. in the context of `USER jovyan`. Generally the pattern to follow is:

1. Do all system installs as root
2. Change to the user that should own files
3. Then copy them in

File: 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/flux.ipynb [hunk 10]
```diff
@@ -1623,7 +1623,7 @@
    "source": [
     "![https://flux-framework.org/flux-operator/_static/images/flux-operator.png](https://flux-framework.org/flux-operator/_static/images/flux-operator.png)\n",
     "\n",
-    "> See you next year! 👋️😎️"
+    "<!-- >> See you next year! 👋️😎️ -->"
```
Comment (vsoch) [inline 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/flux.ipynb:1626]: This is hidden now?

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn (hunk 5 above)
Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn:113]: That makes sense. To be honest, I'm not entirely sure why it's complaining about `$HOME/.local/share` anymore. To my knowledge, nothing is being copied into there.

File: 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/flux.ipynb (hunk 10 above)
Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/flux.ipynb:1626]: Yeah, just a minor thing, but since I'm not sure if we'll be giving this tutorial at JLESC again next year, I didn't think it made sense to keep for this version of the tutorial.

File: 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn (hunk 8 above)
Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/docker/Dockerfile.spawn]: The build on CI worked with this uncommented so it should be fine.

File: 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/flux.ipynb (hunk 10 above)
Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/flux.ipynb:1626]: I also just realized that `dyad.ipynb` is still here. I definitely want to delete that notebook. I'll make a commit for that in a sec

Comment (vsoch) [reply 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/flux.ipynb:1626]: > Actually, @vsoch, I just realized that this change was in flux.ipynb. Part of the plan for this tutorial was to break that down into sections (indicated by the notebooks starting with numbers). All the content from flux.ipynb should already be in those notebooks, so do you think it's fine just to delete flux.ipynb?

I haven't tested it out, so I can't say, but I would rely on your judgment - if you can confidently say nothing is lost it's OK to delete. I'll be running it locally (building now) and I can try to check too, but absence of something is often hard to detect!

Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/flux.ipynb:1626]: That's a good point.

For me, the biggest thing is to not have it seen by the attendees of the tutorial because I feel like that will get confusing. I could always remove it in `Dockerfile.spawn`. That way we still have the file on GitHub, but it won't be there to confuse the attendees.

Comment (ilumsden) [reply 2024-RIKEN-AWS/JupyterNotebook/tutorial/notebook/flux.ipynb:1626]: What do you think, @vsoch?

**Comment by vsoch** [comment]
There is a large organizational issue I see. The first content the user should see when learning about flux is basic job submission. This is what the old tutorial looked like to start.

image

Now, the first flux page is an intro with navigation, and the page that is about submitting jobs goes right into flux batch and interactive jobs. 

image

The `flux submit hostname` easy cases need to come first, and they are (as far as I can tell) entirely gone. For the supplementary section, most of the content is OK there, but the entire chunk of (again submission) in Python, which is a huge use case, is pushed to the end and not properly called out. So my requests:

- Move the original content of submitting jobs, with flux submit, back into that first module.
- Have the flux python submit be a module that comes shortly after that.

Otherwise it's looking beautiful and great! But these things above are big enough that I think will be a hindrance to the user's learning about flux.

**Comment by vsoch** [comment]
I also don't see flux watch showed prominently as it was before.

**Comment by ilumsden** [comment]
The changes regarding the submission APIs were intentional. By far the biggest piece of feedback that we got from people in our dry run on Tuesday was that the old ordering where we dived straight into `flux submit` was confusing. Since they mostly use other batch schedulers, they got lost with us starting with a command with no real equivalent in e.g., Slurm.

This new ordering (starting with `batch` and `alloc`, then moving onto `run` and `submit`) was based on that feedback to be friendly to people with no major background in Flux. The idea was that, in a scheduler like Slurm, users normally get an allocation with e.g., `sbatch` or `salloc` and then launch programs with `srun`. This structure mimics that.

If you want, I could merge that content back into the section on `flux submit`, but I think putting it at the very beginning will cause the confusion that we saw on Tuesday to return.

**Comment by vsoch** [comment]
But the basic submit section is entirely gone - the only `flux submit hostname` is querying for the status. 
image

I don't know who you got feedback from (just your lab? Slurm users?) but I'm not comfortable with the changes to merge it in as a prototype for our tutorials this year - I think I disagree because I find the order (and missing pieces) confusing. I would want to have feedback from flux developers and more folks on our side. Let me think this over for some approach that will let you still run your tutorials without deleting what we have here.

Ping @milroy I need help to think about this.

**Comment by vsoch** [comment]
This is how I see the order - srun is equivalent to submit and run, which are basically the same, but with run you run it "attached" or interactively, and submit just gives you the jobid. I think slurm folds those into one, and for interactive you have to do like `--pty bash` but it's been a hot minute since I used slurm.

**Comment by vsoch** [comment]
I'm going to restore it into a variant I like, push into another branch for the future RADIUSS, and then (given the short time) we can merge what you have here. I need some more time to fix things up - sorry this is really stressful to be doing last minute.

**Comment by vsoch** [comment]
okay, I made a variant (with the initial changes I had and some refactoring) that will ease my mind here, and let us merge / deploy for riken (and not lose what I consider valuable content for RADIUSS) https://github.com/flux-framework/Tutorials/pull/31. As soon as that finishes building OK I should be able to squash here, and then will need to pull and update configs (and likely I'll be doing fixup for the commits so the entire tutorial is just one, there are quite a lot here). Sorry for the delay, just trying to save myself future work.

**Comment by vsoch** [comment]
The one thing I'm not sure about is if we will need the init container - we used it previously to download tutorial files (that are now hosted here) but there was also a change of permissions in the entrypoint. My instinct is saying no, we do not, but also that it might introduce error if I remove them. So I'll remove the clone, but keep the init container.
